src/csharp/Grpc.Core/Internal/AsyncCallBase.cs

```diff
@@ -252,18 +259,39 @@ protected Exception TryDeserialize(byte[] payload, out TRead msg)
         /// </summary>
         protected void HandleSendFinished(bool success)
         {
+            bool delayCompletion = false;
             TaskCompletionSource<object> origTcs = null;
             lock (myLock)
             {
                 origTcs = streamingWriteTcs;
                 streamingWriteTcs = null;
 
+                if (!success && !finished && IsClient)
+                {
+                    // We should be setting this only once per call, following writes will be short circuited.
```

Comment:
They get short circuited later only because the next call can't start until the current one, which waits on the final status, is complete?